rasa/core/domain.py

```diff
@@ -265,30 +265,61 @@ def collect_slots(slot_dict: Dict[Text, Any]) -> List[Slot]:
-                name = list(intent.keys())[0]
-                for properties in intent.values():
-                    properties.setdefault("use_entities", True)
-                    properties.setdefault("ignore_entities", [])
-                    if (
-                        properties["use_entities"] is None
-                        or properties["use_entities"] is False
-                    ):
-                        properties["use_entities"] = []
-            else:
-                name = intent
+            if not isinstance(intent, dict):
                 intent = {intent: {"use_entities": True, "ignore_entities": []}}
 
+            name, properties = list(intent.items())[0]
+
             if name in intent_properties.keys():
                 raise InvalidDomain(
                     "Intents are not unique! Found two intents with name '{}'. "
                     "Either rename or remove one of them.".format(name)
                 )
 
+            if (
+                properties.get("used_entities") is None
+            ):  # important to differentiate between None and [] here
+
+                properties.setdefault("use_entities", True)
+                properties.setdefault("ignore_entities", [])
+                if (
+                    properties["use_entities"] is None
+                    or properties["use_entities"] is False
+                ):
+                    properties["use_entities"] = []
+
+                # `use_entities` is either a list of explicitly included entities
+                # or `True` if all should be included
+                include = properties.get(
+                    "use_entities", True
+                )  # this should even work before setting defaults
+                included_entities = set(entities if include is True else include)
+                excluded_entities = set(properties.get("ignore_entities", []))
+                wanted_entities = list(included_entities - excluded_entities)
+                wanted_entities.sort()
+
+                # Only print warning for ambiguous configurations if entities were included
+                # explicitly.
+                explicitly_included = isinstance(include, list)
+                ambiguous_entities = included_entities.intersection(excluded_entities)
+                if explicitly_included and ambiguous_entities:
+                    raise_warning(
+                        f"Entities: '{ambiguous_entities}' are explicitly included and"
+                        f" excluded for intent '{name}'."
+                        f"Excluding takes precedence in this case. "
+                        f"Please resolve that ambiguity.",
+                        docs=DOCS_URL_DOMAINS
+                        + "#ignoring-entities-for-certain-intents",
+                    )
+
+                properties["used_entities"] = wanted_entities
+                properties.pop("use_entities", None)
+                properties.pop("ignore_entities")
```

Comment:
I think all of this should be its own function as well

Reply:
re this and the previous comment: I'm not sure what functionalities you think are better in separate functions (i.e. starting from where to where). Can you clarify?

Reply:
oh sorry, I thought I'd highlighted a whole block. I meant the whole modification of the `properties` dict, from `if properties.get("used_entities") is None` onwards. Don't worry if it ends up being much more code, this function is just getting a hard to follow